### PR TITLE
Add Face priority  for psi-rules

### PR DESCRIPTION
**Pull request by amebel**
This adds face-priority. By changing the face-priority one is able to modify with which face the next interaction would occur. The face to be interacted next is also a function of the distance from the face being interacted with to the other faces. 

Next integrate face-recognition, tracking and people talking rules.

### CODE REVIEW

File: src/face-priority.scm [hunk 1]
```diff
@@ -0,0 +1,391 @@
...
+  face-id-2.
+
+  A transition priority of 1 means that the system is very likely to shift from
+  face-id-1 to face-id-2 , while a transition priority of 0 means that the
+  system is very unlikely to shift from face-id-1 to face-id-2.
+"
+    (let ((d (distance-in-plane-yz face-id-1 face-id-2)))
+        ; This check is for (inf). If d is not (inf) and is greater than
+        ; the width-of-yz-plane, then something fishy is going on.
+        (if (> d  width-of-yz-plane)
```
Comment (linas) [inline src/face-priority.scm:189]: I am sort-of surprised that you are doing camera and and FOV-plane numeric calculations here.  Ideally, these should really be part of the face-tracker and/or spacetime server subsystem.  That subsystem should provide some API call "distance between faces" and get back some number between 0 and 1.

One fairly serious problem here is that the body-cam on the actual robot is not 640x480, nor is the camera on my desktop and laptop...

**Comment by linas** [comment]
This looks plausible to me and seems OK to merge, although the calculation for how far apart the faces are should really be a service provided by the face-tracker API, rather than a camera geometry calculation you do here.

**Comment by linas** [comment]
At any rate, coordinates should not be in camera pixel coordinates, they should be in degrees or radians or something like that.   Pixels are really dangerous.

File: src/face-priority.scm (hunk 1 above)
Comment (amebel) [reply src/face-priority.scm:189]: I am using a usb camera that is used by the robot from my pc, but I get your valid point.

Comment (amebel) [reply src/face-priority.scm:189]: Just found out that it isn't 640x480, but is scaled down or something like that, not totally clear how/why though.

**Comment by amebel** [comment]
Ok will move the face distance calculation to face_track.

The coordinate isn't in pixels it is in meters. cmt and the pi_vision use the formula @ https://github.com/hansonrobotics/HEAD/blob/master/src/vision/pi_vision/pi_face_tracker/nodes/face_tracker.py#L271 . I just used it to project the coordinates to a plane at 1 meter distance from the camera. 

I suppose using degrees/radians for face position would make the calculations camera neutral, as long as depth isn't required.

**Comment by linas** [comment]
Meters is OK. I worked hard to make sure it was meters. 

If, instead of projecting to a _plane_ one meter from the camera, you projected to a _sphere_ one meter from the camera, then the width of one meter is _exactly_ one radian. (by definition) Recall that there are two-pi radians in a circle, no matter how big the circle.

**Comment by amebel** [comment]
I don't get how I projected to a sphere, what am I missing?

**Comment by linas** [comment]
I was suggesting that you _should_ project onto a sphere, instead of a plane, and use radians, not pixels or meters, as the natural unit of measure. 

However, perhaps it doesn't matter so much -- what's being done here is sufficiently crude that the difference between a plane projection and a sphere projection probably wouldn't change any results. It still would be nice to nuke the 640x480 thing, though.  That's like an unexploded land-mine, waiting for someone to step on it.

**Pull request merged**